File: functions/src/quranSessions/stagingQaJoinWindowBypass.ts
import * as logger from "firebase-functions/logger";

import { FIREBASE_PROJECT_ID } from "../github";
import {
  MAESTRO_STUDENT_UID,
  MAESTRO_TEACHER_UID,
} from "./maestroStagingAccounts";

/** Staging Maestro QA accounts — join-window bypass only, never production. */
export const STAGING_QA_JOIN_WINDOW_BYPASS_UIDS = new Set<string>([
  MAESTRO_TEACHER_UID,
  MAESTRO_STUDENT_UID,
]);

const BLOCKED_DISTRIBUTIONS = new Set(["production", "play_production"]);

const STAGING_DISTRIBUTIONS = new Set(["local", "staging"]);

/**
 * True when Cloud Functions run in a non-production staging context.
 *
 * Prefers explicit `TILAWA_DISTRIBUTION`; falls back to Firebase project id
 * (`quran-playera-app` staging project).
 */
export function isStagingEnvironmentForQaJoinWindowBypass(): boolean {
  const distribution = process.env.TILAWA_DISTRIBUTION?.trim();
  if (distribution != null && BLOCKED_DISTRIBUTIONS.has(distribution)) {
    return false;
  }
  if (distribution != null && STAGING_DISTRIBUTIONS.has(distribution)) {
    return true;
  }

  const projectId =
    process.env.GCLOUD_PROJECT?.trim()
    ?? process.env.FIREBASE_PROJECT_ID?.trim()
    ?? FIREBASE_PROJECT_ID;
  return projectId === FIREBASE_PROJECT_ID;
}

/**
 * Staging-only QA override: skip join-window timing for allowlisted uids.
 * All other sessionAuth / lifecycle / participant checks remain enforced.
 */
export function isQaJoinWindowBypassEligible(uid: string | undefined): boolean {
  const normalized = uid?.trim();
  if (normalized == null || normalized.length === 0) {
    return false;
  }
  if (!STAGING_QA_JOIN_WINDOW_BYPASS_UIDS.has(normalized)) {
    return false;
  }
  if (!isStagingEnvironmentForQaJoinWindowBypass()) {
    return false;
  }

  logger.info(`[QA] join-window bypass applied for uid=${normalized}`);
  return true;
}

File: packages/quran_sessions/lib/src/domain/policies/staging_qa_join_window_bypass.dart
import 'dart:developer' as developer;

/// Staging Maestro QA teacher uid (`mu7ammadkamel@hotmail.com`).
const stagingQaTeacherUid = 'WV0m6tenTJPDLZE4EdWXBzjADF12';

/// Staging Maestro QA student uid (`mohammad.kamel@othaimmarkets.com`).
const stagingQaStudentUid = 'U33e4w08bYWFOuS7NTxoHmvDFxM2';

/// Staging-only uids that may bypass join-window timing (QA / Maestro).
const stagingQaJoinWindowBypassUids = <String>{
  stagingQaTeacherUid,
  stagingQaStudentUid,
};

const _blockedDistributions = {'production', 'play_production'};

const _stagingDistributions = {'local', 'staging'};

/// Compile-time distribution from `--dart-define=TILAWA_DISTRIBUTION=…`.
const stagingQaJoinWindowBypassDefaultDistribution = String.fromEnvironment(
  'TILAWA_DISTRIBUTION',
  defaultValue: 'local',
);

bool isStagingDistributionForQaJoinWindowBypass([
  String distribution = stagingQaJoinWindowBypassDefaultDistribution,
]) {
  if (_blockedDistributions.contains(distribution)) {
    return false;
  }
  return _stagingDistributions.contains(distribution);
}

/// Staging-only QA override: skip join-window timing for allowlisted uids.
///
/// Lifecycle, participant, and terminal-state checks stay enforced elsewhere.
bool isQaJoinWindowBypassEligible({
  required String? userId,
  String distribution = stagingQaJoinWindowBypassDefaultDistribution,
}) {
  final normalized = userId?.trim();
  if (normalized == null || normalized.isEmpty) {
    return false;
  }
  if (!stagingQaJoinWindowBypassUids.contains(normalized)) {
    return false;
  }
  if (!isStagingDistributionForQaJoinWindowBypass(distribution)) {
    return false;
  }

  developer.log(
    '[QA] join-window bypass applied for uid=$normalized',
    name: 'quran_sessions.qa_join_window',
  );
  return true;
}
